replace a few do-nothing destructors by default

nschloe opened the pull request:
In C++11, special member functions (such as the move constructor) are only
created by the compiler is none of destructor, copy constructor, and copy
assignment operator are defined; see, e.g.,
<http://stackoverflow.com/a/4782927/353337>.

This PR removes a few destructors such that move constructors are automatically
generated.

Caught by Coverity <https://scan.coverity.com/projects/mixxxdj-mixxx>.

nschloe committed:
remove a few do-nothing destructors

In C++11, special member functions (such as the move constructor) are only
created by the compiler is none of destructor, copy constructor, and copy
assignment operator are defined; see, e.g.,
<http://stackoverflow.com/a/4782927/353337>.

This PR removes a few destructors such that move constructors are automatically
generated.

Caught by Coverity <https://scan.coverity.com/projects/mixxxdj-mixxx>.

uklotzde commented on src/util/audiosignal.h:
```diff
@@ -70,7 +70,6 @@ class AudioSignal {
         DEBUG_ASSERT(kChannelCountZero <= m_channelCount); // unsigned value
         DEBUG_ASSERT(kSamplingRateZero <= m_samplingRate); // unsigned value
     }
-    virtual ~AudioSignal() {}
```
C.35: A base class destructor should be either public and virtual, or protected and nonvirtual
http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#a-namerc-dtor-virtualac35-a-base-class-destructor-should-be-either-public-and-virtual-or-protected-and-nonvirtual

nschloe replied:
Hm, that's right! I'll `=default` a bunch of functions then.

uklotzde commented on src/skin/pixmapsource.cpp:
```diff
@@ -10,9 +10,6 @@ PixmapSource::PixmapSource(const QString& filepath) {
     setPath(filepath);
 }
 
-PixmapSource::~PixmapSource() {
```
class PixmapSource should be declared final without a virtual destructor

uklotzde commented on src/library/coverart.h:
```diff
@@ -30,7 +30,6 @@ class CoverInfoRelative {
     };
 
     CoverInfoRelative();
-    virtual ~CoverInfoRelative() {};
```
CoverInfoRelative is a base class (see remark below)

uklotzde commented on src/effects/effectmanifest.h:
```diff
@@ -27,9 +27,6 @@ class EffectManifest {
           m_isForFilterKnob(false),
           m_effectRampsFromDry(false) {
     }
-    virtual ~EffectManifest() {
```
class EffectManifest should be declared _final_ without a virtual destructor

uklotzde commented on src/library/coverart.h:
```diff
@@ -51,8 +50,6 @@ class CoverInfo : public CoverInfoRelative {
           trackLocation(tl) {
     }
 
-    virtual ~CoverInfo() {};
```
I prefer to define a destructor in derived classes even if it is empty. By declaring it as _override_ instead of _virtual_ the compiler checks that the base class indeed has a virtual destructor.

uklotzde commented on src/library/coverart.h:
```diff
@@ -72,8 +69,6 @@ class CoverArt : public CoverInfo {
           resizedToWidth(rtw) {
     }
 
-    virtual ~CoverArt() {};
```
Same here, override virtual destructor of base class.

uklotzde commented:
+1 defaulting is even better than an empty implementation! Good idea.

On 02.01.2017 11:47, Nico Schlömer wrote:
>
> *@nschloe* commented on this pull request.
>
> ----------------------------------------------------------------------------
>
> In src/util/audiosignal.h <https://github.com/mixxxdj/mixxx/pull/1100>:
>
> > @@ -70,7 +70,6 @@ class AudioSignal {
>           DEBUG_ASSERT(kChannelCountZero <= m_channelCount); // unsigned value
>           DEBUG_ASSERT(kSamplingRateZero <= m_samplingRate); // unsigned value
>       }
> -    virtual ~AudioSignal() {}
>
> Hm, that's right! I'll |=default| a bunch of functions then.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/1100>, or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ABZ-as7D_lc9rp1tQqq23N1qEHvqm7S3ks5rONXSgaJpZM4LYkGZ>.
>

nschloe committed:
be more explicit with the default special member functions

nschloe commented:
@uklotzde Done.

rryan commented:
```
cl /Fowin64_build\plugins\win64_build\mediafoundation\soundsourcemediafoundation.obj /c plugins\soundsourcemediafoundation\soundsourcemediafoundation.cpp /TP /nologo /MP /Zi /MD /GL /fp:fast /Gy /O2 /Damd64 /DMIXXX_BUILD_RELEASE /D'Q_ASSERT(x)=qt_noop()' /D__WINDOWS__ /D_ATL_XP_TARGETING /D_ATL_MIN_CRT /DNOMINMAX /DUNICODE /DWIN64 /D_USE_MATH_DEFINES /DSETTINGS_FILE=\"mixxx.cfg\" /D__PORTAUDIO__ /DQT_TABLET_SUPPORT /DQT_SHARED /DQT_CORE_LIB /DQT_GUI_LIB /DQT_OPENGL_LIB /DQT_XML_LIB /DQT_SVG_LIB /DQT_SQL_LIB /DQT_SCRIPT_LIB /DQT_NETWORK_LIB /D__SNDFILE__ /DPROTOBUF_USE_DLLS /D__MAD__ /D__MEDIAFOUNDATION__ /D__HSS1394__ /D__HID__ /D__VINYLCONTROL__ /D__BROADCAST__ /D__OPUS__ /D__SSE__ /D__SSE2__ /D__SSE__ /D__SSE2__ /D__VAMP__ /Dkiss_fft_scalar=double /D__SQLITE3__ /D__BATTERY__ /D_USING_V110_SDK71_ /IC:\mixxx-buildserver\2.0-x64-release-minimal /IC:\mixxx-buildserver\2.0-x64-release-minimal\include /Iwin64_build\plugins\win64_build\mediafoundation /Iplugins\soundsourcemediafoundation /Isrc /Iplugins\win64_build\mediafoundation /Isrc\plugins\win64_build\mediafoundation /Ilib\soundtouch-1.9.2 /Ilib\replaygain /Ilib\libebur128-1.1.0\ebur128 /Ilib\libebur128-1.1.0\ebur128\queue /IC:\mixxx-buildserver\2.0-x64-release-minimal\build\qt-everywhere-opensource-src-4.8.6\include /IC:\mixxx-buildserver\2.0-x64-release-minimal\build\qt-everywhere-opensource-src-4.8.6\include /IC:\mixxx-buildserver\2.0-x64-release-minimal\build\qt-everywhere-opensource-src-4.8.6\include\QtCore /IC:\mixxx-buildserver\2.0-x64-release-minimal\build\qt-everywhere-opensource-src-4.8.6\include\QtGui /IC:\mixxx-buildserver\2.0-x64-release-minimal\build\qt-everywhere-opensource-src-4.8.6\include\QtOpenGL /IC:\mixxx-buildserver\2.0-x64-release-minimal\build\qt-everywhere-opensource-src-4.8.6\include\QtXml /IC:\mixxx-buildserver\2.0-x64-release-minimal\build\qt-everywhere-opensource-src-4.8.6\include\QtSvg /IC:\mixxx-buildserver\2.0-x64-release-minimal\build\qt-everywhere-opensource-src-4.8.6\include\QtSql /IC:\mixxx-buildserver\2.0-x64-release-minimal\build\qt-everywhere-opensource-src-4.8.6\include\QtScript /IC:\mixxx-buildserver\2.0-x64-release-minimal\build\qt-everywhere-opensource-src-4.8.6\include\QtXmlPatterns /IC:\mixxx-buildserver\2.0-x64-release-minimal\build\qt-everywhere-opensource-src-4.8.6\include\QtNetwork /IC:\mixxx-buildserver\2.0-x64-release-minimal\build\qt-everywhere-opensource-src-4.8.6\include\QtTest /IC:\mixxx-buildserver\2.0-x64-release-minimal\build\qt-everywhere-opensource-src-4.8.6\include\QtScriptTools /Ilib\gtest-1.7.0\include /Ilib\fidlib-0.9.10 /IC:\usr\include\taglib /Ilib\qtscript-bytearray /Ilib\reverb /Ilib\portaudio /Ilib\hidapi-0.8.0-rc1\hidapi /Ilib\xwax /Ilib\scratchlib /Ilib\vamp-2.6
cl @c:\users\appveyor\appdata\local\temp\1\tmpmtlctn.lnk
soundsourcemediafoundation.cpp
C:\projects\mixxx\src\util/audiosignal.h(76) : error C2610: 'mixxx::AudioSignal::AudioSignal(mixxx::AudioSignal &&)' : is not a special member function which can be defaulted
C:\projects\mixxx\src\util/audiosignal.h(78) : error C2610: 'mixxx::AudioSignal &mixxx::AudioSignal::operator =(mixxx::AudioSignal &&)' : is not a special member function which can be defaulted
```

nschloe commented:
@rryan Uh, never saw this error. Looking at `C:\`, I suppose this is on Windows. Looks like the compiler there doesn't know move semantics. What compiler and version are used here?

rryan commented:
> @rryan Uh, never saw this error. Looking at C:\, I suppose this is on Windows. Looks like the compiler there doesn't know move semantics. What compiler and version are used here?

Yes, this is Windows. You can see the results in the Appveyor section of the build checks.

We use MSVC 2013 at the moment, and may upgrade to 2015 for Mixxx 2.1 or 2.2.

uklotzde commented:
Yes. I remember that VS2013 doesn't support 'default' for move constructors. 
We should definitely raise the build requirements ASAP.

On 02.01.2017 18:52, Nico Schlömer wrote:
>
> @rryan <https://github.com/rryan> Uh, never saw this error. Looking at 
> |C:\|, I suppose this is on Windows. Looks like the compiler there doesn't 
> know move semantics. What compiler and version are used here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/1100#issuecomment-270002059>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ABZ-akp9euy_hxNKj-GjVVpqH9IrXqAeks5rOTlQgaJpZM4LYkGZ>.
>

nschloe commented:
Aha.

I'm reading [here](http://stackoverflow.com/a/26581337/353337):

> "Rvalue references v3.0" adds new rules to automatically generate move constructors and move assignment operators under certain conditions. However, this is not implemented in Visual C++ in Visual Studio 2013, due to time and resource constraints.

[Apparently](https://msdn.microsoft.com/en-us/library/hh567368.aspx), this has been fixed in MSVC 2015:

> "Rvalue references v3.0" adds new rules to automatically generate move constructors and move assignment operators under certain conditions. This is implemented in Visual Studio 2015.

uklotzde commented:
https://msdn.microsoft.com/en-us/library/hh567368.aspx*

*Defaulted and Deleted Functions
These are now supported, but with this exception: For defaulted functions, 
the use of =default to request member-wise move constructors and move 
assignment operators is not supported.
*
*On 02.01.2017 18:56, RJ Skerry-Ryan wrote:
>
>     @rryan <https://github.com/rryan> Uh, never saw this error. Looking at
>     C:, I suppose this is on Windows. Looks like the compiler there
>     doesn't know move semantics. What compiler and version are used here?
>
> Yes, this is Windows. You can see the results in the Appveyor section of 
> the build checks.
>
> We use MSVC 2013 at the moment, and may upgrade to 2015 for Mixxx 2.1 or 2.2.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/1100#issuecomment-270002547>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ABZ-aj-yHCmxO_Yl-BBD5431SczaY19gks5rOTphgaJpZM4LYkGZ>.
>

nschloe commented:
Ugh. :/

Okay, I see three possibilities here:

 * `#ifdef` the `=default` statements. Adds a bunch of lines.
 * remove all `=default` blocks and have them generated by the compiler. This disregards @uklotzde's comments from above (about being explicit about the special member functions).
 * Forget this PR and move semantics.

rryan commented:
MSVC15 plus our static build setup is currently very unstable so we can't switch quite yet. :(
https://bugs.launchpad.net/bugs/1653368

uklotzde commented:
Even VS2015 does not support default move constructors.

I prefer conditional compilation to explicitly document this deficiency:
```
#if defined(_MSC_VER) && (_MSC_VER <= 1900)
    // Visual Studio 2015 does not support default generated move constructors
    TrackCacheResolver(TrackCacheResolver&& moveable)
        : TrackCacheLocker(std::move(moveable)) {
    }
#else
    TrackCacheResolver(TrackCacheResolver&&) = default;
#endif
```

nschloe committed:
ifdef out default move function for MSVC

From <https://msdn.microsoft.com/en-us/library/hh567368.aspx>:

Defaulted and Deleted Functions
These are now supported, but with this exception: For defaulted functions, the
use of =default to request member-wise move constructors and move assignment
operators is not supported.

nschloe commented:
@rryan Could you give it another shot?

rryan commented:
AppVeyor passed, so this means windows 32-bit and 64-bit compiled and tests passed. Good enough for me!

rryan commented:
thanks @nschloe !

Pegasus-RPG commented on src/library/coverart.h:
```diff
@@ -33,10 +33,13 @@ class CoverInfoRelative {
-    CoverInfoRelative& operator=(CoverInfoRelative&&) = default;
     virtual ~CoverInfoRelative() = default;
+// Visual Studio does not support default generated move constructors
+#if !defined(_MSC_VER)
```
This should include the version number since (hopefully) future versions of MSVC will work.

uklotzde replied:
See my example above. I'm enabling the workaround only for versions that are known **not** to work, i.e. everything including VS2015 -> (_MSC_VER <= 1900). For VS2017 the workaround is disabled and we will notice early if it works or not.

nschloe replied:
>  For VS2017 the workaround is disabled and we will notice early if it works or not.

If you're up for compile failures in the future, I'll be happy to include that version statement.

uklotzde commented:
I prefer to be more restrictive. The compile time failiure indicates that 
you need to revisit this code. It's like a reminder ;)

The same approach works for runtime assertions. Be as restrictive as 
possible! They are only evaluated during development anyway. Whenever such 
an assertion fails, verify your assumptions and the code again. Either your 
assumption or the code might be wrong.

On 04.01.2017 08:51, Nico Schlömer wrote:
>
> *@nschloe* commented on this pull request.
>
> ----------------------------------------------------------------------------
>
> In src/library/coverart.h <https://github.com/mixxxdj/mixxx/pull/1100>:
>
> >      virtual ~CoverInfoRelative() = default;
> +// Visual Studio does not support default generated move constructors
> +#if !defined(_MSC_VER)
>
>     For VS2017 the workaround is disabled and we will notice early if it
>     works or not.
>
> If you're up for compile failures in the future, I'll be happy to include 
> that version statement.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/1100>, or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ABZ-at_5cOJYXfZ_q2Y5_l9arxArgaaVks5rO0-NgaJpZM4LYkGZ>.
>

nschloe commented:
See PR #1107.
